state/lastSnapshotMarker/lastSnapshotMarker.go

```diff
@@ -0,0 +1,82 @@
+package lastSnapshotMarker
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-go/common"
+	"github.com/multiversx/mx-chain-go/common/storage"
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("state/lastSnapshotMarker")
+
+const (
+	lastSnapshot = "lastSnapshot"
+)
+
+type lastSnapshotMarker struct {
+	mutex                       sync.RWMutex
+	latestFinishedSnapshotEpoch uint32
+}
+
+func NewLastSnapshotMarker() *lastSnapshotMarker {
+	return &lastSnapshotMarker{}
+}
+
+func (lsm *lastSnapshotMarker) AddMarker(trieStorageManager common.StorageManager, epoch uint32, rootHash []byte) {
+	err := storage.WaitForStorageEpochChange(storage.StorageEpochChangeWaitArgs{
+		TrieStorageManager:            trieStorageManager,
+		Epoch:                         epoch,
+		WaitTimeForSnapshotEpochCheck: storage.WaitTimeForSnapshotEpochCheck,
+		SnapshotWaitTimeout:           storage.SnapshotWaitTimeout,
+	})
+	if err != nil {
+		log.Warn("err while waiting for storage epoch change", "err", err, "epoch", epoch, "rootHash", rootHash)
+		return
+	}
+
+	lsm.mutex.Lock()
+	defer lsm.mutex.Unlock()
+
+	if epoch <= lsm.latestFinishedSnapshotEpoch {
+		log.Debug("will not put lastSnapshot marker in epoch storage",
+			"epoch", epoch,
+			"latestFinishedSnapshotEpoch", lsm.latestFinishedSnapshotEpoch,
+		)
+		return
+	}
+
+	err = trieStorageManager.PutInEpoch([]byte(lastSnapshot), rootHash, epoch)
+	if err != nil {
+		log.Warn("could not set lastSnapshot", err, "rootHash", rootHash, "epoch", epoch, "rootHash", rootHash)
+	}
+}
+
+func (lsm *lastSnapshotMarker) RemoveMarker(trieStorageManager common.StorageManager, epoch uint32, rootHash []byte) {
+	lsm.mutex.RLock()
```

Comment:
Because of L64 I think this should have been a Lock. Also. The mutex is used here as a critical section or just to protect the `latestFinishedSnapshotEpoch` value?

Reply:
Indeed, changed to `Lock()` and `Unlock()` . Critical section needed here.